File: elevator.c
#include <stdio.h>
#include <unistd.h> // for sleep function

// Enum to represent the current state of the elevator
enum ElevatorState {
    IDLE,
    UP,
    DOWN
};

// Function to move the elevator up
void moveUp(int destinationFloor) {
    for (int currentFloor = 0; currentFloor < destinationFloor; ++currentFloor) {
        printf("Elevator at floor %d.\n", currentFloor);

        // Check for emergency interrupt
        char response;
        printf("If emergency, press 'S' to stop. Continue with 'G': ");
        scanf(" %c", &response);

        if (response == 'S') {
            printf("Emergency stop. Elevator paused.\n");
            printf("Press 'G' to resume.\n");
            
            // Wait for 'G' to resume
            while (1) {
                scanf(" %c", &response);
                if (response == 'G') {
                    break;
                }
            }

            printf("Resuming elevator.\n");
        }

        sleep(1); // 1-second delay between floors
    }
    printf("Elevator arrived at floor %d.\n", destinationFloor);

    // Open and close the elevator door
    printf("Press 'C' to open the door: ");
    scanf(" %c", &response);
    if (response == 'C') {
        printf("Elevator door opening.\n");
        sleep(3); // 3-second delay for the door to stay open
        printf("Elevator door closing.\n");
    }

    // Add code to control the elevator motor and update state
}

// Function to move the elevator down
void moveDown(int destinationFloor) {
    for (int currentFloor = 0; currentFloor > destinationFloor; --currentFloor) {
        printf("Elevator at floor %d.\n", currentFloor);

        // Check for emergency interrupt
        char response;
        printf("If emergency, press 'S' to stop. Continue with 'G': ");
        scanf(" %c", &response);

        if (response == 'S') {
            printf("Emergency stop. Elevator paused.\n");
            printf("Press 'G' to resume.\n");
            
            // Wait for 'G' to resume
            while (1) {
                scanf(" %c", &response);
                if (response == 'G') {
                    break;
                }
            }

            printf("Resuming elevator.\n");
        }

        sleep(1); // 1-second delay between floors
    }
    printf("Elevator arrived at floor %d.\n", destinationFloor);

    // Open and close the elevator door
    printf("Press 'C' to open the door: ");
    scanf(" %c", &response);
    if (response == 'C') {
        printf("Elevator door opening.\n");
        sleep(3); // 3-second delay for the door to stay open
        printf("Elevator door closing.\n");
    }

    // Add code to control the elevator motor and update state
}

int main() {
    enum ElevatorState currentState = IDLE;
    int destinationFloor;

    while (1) {
        // Display the menu
        printf("\nElevator Control Menu:\n");
        printf("1. Move Up\n");
        printf("2. Move Down\n");
        printf("0. Exit\n");
        printf("Enter your choice: ");

        int choice;
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter the desired floor (between basement and level 50): ");
                scanf("%d", &destinationFloor);

                // Validate the input floor range
                if (destinationFloor < -1 || destinationFloor > 50) {
                    printf("Invalid floor selection. Please choose a floor between basement and level 50.\n");
                    continue; // Skip to the next iteration of the loop
                }

                // Request to move the elevator up
                if (currentState == IDLE) {
                    moveUp(destinationFloor);
                    currentState = UP;
                }
                break;

            case 2:
                printf("Enter the desired floor (between basement and level 50): ");
                scanf("%d", &destinationFloor);

                // Validate the input floor range
                if (destinationFloor < -1 || destinationFloor > 50) {
                    printf("Invalid floor selection. Please choose a floor between basement and level 50.\n");
                    continue; // Skip to the next iteration of the loop
                }

                // Request to move the elevator down
                if (currentState == IDLE) {
                    moveDown(destinationFloor);
                    currentState = DOWN;
                }
                break;

            case 0:
                printf("Exiting elevator control program. Goodbye!\n");
                return 0;

            default:
                printf("Invalid choice. Please enter a valid option.\n");
        }
    }

    return 0;
}

File: elevator.java
import java.util.Scanner;

// Enum to represent the current state of the elevator
enum ElevatorState {
    IDLE,
    UP,
    DOWN
}

public class ElevatorControl {
    public static void main(String[] args) {
        ElevatorState currentState = ElevatorState.IDLE;
        int destinationFloor;

        Scanner scanner = new Scanner(System.in);

        while (true) {
            // Display the menu
            System.out.println("\nElevator Control Menu:");
            System.out.println("1. Move Up");
            System.out.println("2. Move Down");
            System.out.println("0. Exit");
            System.out.print("Enter your choice: ");

            int choice = scanner.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Enter the desired floor (between basement and level 50): ");
                    destinationFloor = scanner.nextInt();

                    // Validate the input floor range
                    if (destinationFloor < -1 || destinationFloor > 50) {
                        System.out.println("Invalid floor selection. Please choose a floor between basement and level 50.");
                        continue; // Skip to the next iteration of the loop
                    }

                    // Request to move the elevator up
                    if (currentState == ElevatorState.IDLE) {
                        moveUp(destinationFloor);
                        currentState = ElevatorState.UP;
                    }
                    break;

                case 2:
                    System.out.print("Enter the desired floor (between basement and level 50): ");
                    destinationFloor = scanner.nextInt();

                    // Validate the input floor range
                    if (destinationFloor < -1 || destinationFloor > 50) {
                        System.out.println("Invalid floor selection. Please choose a floor between basement and level 50.");
                        continue; // Skip to the next iteration of the loop
                    }

                    // Request to move the elevator down
                    if (currentState == ElevatorState.IDLE) {
                        moveDown(destinationFloor);
                        currentState = ElevatorState.DOWN;
                    }
                    break;

                case 0:
                    System.out.println("Exiting elevator control program. Goodbye!");
                    scanner.close();
                    System.exit(0);

                default:
                    System.out.println("Invalid choice. Please enter a valid option.");
            }
        }
    }

    // Function to move the elevator up
    private static void moveUp(int destinationFloor) {
        for (int currentFloor = 0; currentFloor < destinationFloor; ++currentFloor) {
            System.out.println("Elevator at floor " + currentFloor);
            // Add code to control the elevator motor and update state

            // Simulate 1-second delay between floors
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        System.out.println("Elevator arrived at floor " + destinationFloor);

        // Open and close the elevator door
        System.out.print("Press 'C' to open the door: ");
        Scanner scanner = new Scanner(System.in);
        char response = scanner.next().charAt(0);
        if (response == 'C') {
            System.out.println("Elevator door opening.");
            // Simulate 3-second delay for the door to stay open
            try {
                Thread.sleep(3000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            System.out.println("Elevator door closing.");
        }

        // Add code to control the elevator motor and update state
    }

    // Function to move the elevator down
    private static void moveDown(int destinationFloor) {
        for (int currentFloor = 0; currentFloor > destinationFloor; --currentFloor) {
            System.out.println("Elevator at floor " + currentFloor);
            // Add code to control the elevator motor and update state

            // Simulate 1-second delay between floors
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        System.out.println("Elevator arrived at floor " + destinationFloor);

        // Open and close the elevator door
        System.out.print("Press 'C' to open the door: ");
        Scanner scanner = new Scanner(System.in);
        char response = scanner.next().charAt(0);
        if (response == 'C') {
            System.out.println("Elevator door opening.");
            // Simulate 3-second delay for the door to stay open
            try {
                Thread.sleep(3000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            System.out.println("Elevator door closing.");
        }

        // Add code to control the elevator motor and update state
    }
}
